Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -1928,6 +1929,71 @@ def __setstate_cython__(self, __pyx_state):
             self.exit_scope()
             node.body.stats.append(pickle_func)
 
+    def _inject_init_subclass(self, node):
+        """
+        Performs a bit of validation when cdef classes are inherited from
+        to diagnose issues __cinit__ and __dealloc__ won't be called.
+        This can happen with multiple inheritance from (layout compatible)
+        cdef classes.
+
+        The validation only happens reliably on Python 3.6+, but that's OK because
+        it's a "nice to have" warning for an odd corner-case only. (In some cases
+        Cython itself calls __init_subclass__ on earlier versions though)
+        """
+        if node.scope.lookup('__init_subclass__'):
+            return  # shouldn't happen currently because it's not supported for extension types
+                    # (but may in future)
+
+        # create a cutdown pipeline that stops at AnalyseDeclarationsTransform and excludes
+        # some functions that break and that don't seem to be necessary for successful compilation
+        from .Pipeline import create_pipeline
+        orig_pipeline = create_pipeline(self.context, "pyx")
+        pipeline = []
+        for element in orig_pipeline:
+            if isinstance(element, (PostParse, ParallelRangeTransform, AdjustDefByDirectives,
+                                    AutoCpdefFunctionDefinitions, RemoveUnreachableCode,
+                                    DecoratorTransform)):
+                continue
+            elif isinstance(element, AnalyseDeclarationsTransform):
+                break
+            pipeline.append(element)
+
+        func = TreeFragment(u"""
+            @classmethod
+            def __init_subclass__(cls, **kwds):
+                # it'd be nice to use "issubclass" but it isn't available in TreeFragment
+                try:
+                    all_bases = cls.__base__.__mro__
+                except AttributeError:
+                    # Python 2 workaround - remove me eventually!
+                    def get_bases(C):
```

Comment:
I think `get_bases` is worth factoring out into some common utility code - it causes some quite substantial chunks of code to be generated for every cdef class.